### PR TITLE
Document `stream()` detection limitation

### PR DESCRIPTION
**Pull request by Borewit**
Changes:
1. Document `stream()` limitation.
1. Make sample size in `stream()` configurable, and add corresponding unit tests

Includes PR: #468 (better to merge that one first)

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -254,6 +254,10 @@ Detect the file type of a readable stream.
 
 Returns a `Promise` which resolves to the original readable stream argument, but with an added `fileType` property, which is an object like the one returned from `FileType.fromFile()`.
 
+TVery handy to put in between a stream, but it comes for a price.
```
Comment (sindresorhus) [inline readme.md:257]: Typo

**Comment by sindresorhus** [comment]
Could we maybe let the user decide how big of a sample they want? That way they could decide whether they favor performance or accuracy.

File: core.js [hunk 2]
```diff
@@ -1414,9 +1414,10 @@ async function _fromTokenizer(tokenizer) {
 	}
 }
 
-const stream = readableStream => new Promise((resolve, reject) => {
+const stream = (readableStream, sampleSize) => new Promise((resolve, reject) => {
```
Comment (sindresorhus) [inline core.js:1417]: This should be an options-object. Reading `stream(someStream, 100)` at the callsite is not very dev friendly.

File: readme.md [hunk 3]
```diff
@@ -278,12 +278,16 @@ Type: [`ITokenizer`](https://github.com/Borewit/strtok3#tokenizer)
...
-### FileType.stream(readableStream)
+### FileType.stream(readableStream, sampleSize)
 
 Detect the file type of a readable stream.
 
+If `sampleSize` is not provided, a backward-compatible sample size of 4100 bytes is used.
+
 Returns a `Promise` which resolves to the original readable stream argument, but with an added `fileType` property, which is an object like the one returned from `FileType.fromFile()`.
 
+Very handy to put in between a stream, but it comes with a price. Internally `stream()` builds up a buffer of `sampleSize` bytes, used as a sample, to determine the file type. The sample size impacts the file detection resolution. A smaller sample size will result in lower probability of the best file type detection.
```
Comment (sindresorhus) [inline readme.md:289]:
```suggestion
This method can be handy to put in between a stream, but it comes with a price. Internally, `stream()` builds up a buffer of `sampleSize` bytes, used as a sample, to determine the file type. The sample size impacts the file detection resolution. A smaller sample size will result in lower probability of the best file type detection.
```

?

Comment (Borewit) [reply readme.md:289]: Fully agree

**Comment by Borewit** [comment]
I should actually put a bit of unit testing in place, with an exemplary pointing out (& testing one) with a different sample size.

File: readme.md [hunk 4]
```diff
@@ -278,12 +278,18 @@ Type: [`ITokenizer`](https://github.com/Borewit/strtok3#tokenizer)
 
 A file source implementing the [tokenizer interface](https://github.com/Borewit/strtok3#tokenizer).
 
-### FileType.stream(readableStream)
+### FileType.stream(readableStream, sampleSize)
```
Comment (sindresorhus) [inline readme.md:281]: This needs to be properly documented as an options-object. Example: https://github.com/sindresorhus/boxen#options

File: core.d.ts [hunk 5]
```diff
@@ -355,10 +355,21 @@ declare namespace core {
 	*/
 	const mimeTypes: Set<core.MimeType>;
 
+	/**
+	 * Stream options.
+	 */
+	interface IStreamOptions {
```
Comment (sindresorhus) [inline core.d.ts:361]:
```suggestion
	interface StreamOptions {
```

File: core.d.ts [hunk 6]
```diff
@@ -355,10 +355,21 @@ declare namespace core {
...
+		/**
+		 *  Sample size in bytes.
```
Comment (sindresorhus) [inline core.d.ts:363]: It should document the default value using standard docs annotation.

File: core.d.ts [hunk 7]
```diff
@@ -355,10 +355,21 @@ declare namespace core {
 	*/
 	const mimeTypes: Set<core.MimeType>;
 
+	/**
+	 * Stream options.
+	 */
```
Comment (sindresorhus) [inline core.d.ts:360]: This is moot. It just says what the type name already says.

File: readme.md [hunk 8]
```diff
@@ -278,12 +278,18 @@ Type: [`ITokenizer`](https://github.com/Borewit/strtok3#tokenizer)
...
 
 Detect the file type of a readable stream.
 
+If `sampleSize` is not provided, a backward compatible sample size of 4100 bytes is used.
+
 Returns a `Promise` which resolves to the original readable stream argument, but with an added `fileType` property, which is an object like the one returned from `FileType.fromFile()`.
 
+This method can be handy to put in between a stream, but it comes with a price.
+Internally `stream()` builds up a buffer of `sampleSize` bytes, used as a sample, to determine the file type.
+The sample size impacts the file detection resolution. A smaller sample size will result in lower probability of the best file type detection.
```
Comment (sindresorhus) [inline readme.md:291]: Readme and index.d.ts should be in sync.

File: test.js [hunk 9]
```diff
@@ -351,6 +357,16 @@ test('.stream() method - error event', async t => {
 	await t.throwsAsync(FileType.stream(readableStream), errorMessage);
 });
 
+test('.stream() method - sampleSize', async t => {
```
Comment (sindresorhus) [inline test.js:360]:
```suggestion
test('.stream() method - sampleSize option', async t => {
```

File: test.js [hunk 10]
```diff
@@ -337,6 +337,12 @@ test('.stream() method - short stream', async t => {
 	t.deepEqual(bufferA, bufferB);
 });
 
+test('.stream() method - no End-Of-Stream errors', async t => {
```
Comment (sindresorhus) [inline test.js:340]:
```suggestion
test('.stream() method - no end-of-stream errors', async t => {
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by Borewit** [comment]
Note that the example provided for `stream()` is not working, see issue #470.

**Comment by sindresorhus** [comment]
Not done: https://github.com/sindresorhus/file-type/pull/434#discussion_r674708584

File: core.d.ts (hunk 6 above)
Comment (Borewit) [reply core.d.ts]: > Not done: [#434 (comment)](https://github.com/sindresorhus/file-type/pull/434#discussion_r674708584)

Fixed

**Pull request merged**